### ISSUE TITLE
Possible to use normal SPI instead of QSPI?

### ISSUE THREAD

**Issue by gampam2000** [nspsck/RM67162_Micropython_QSPI#1]
Hi,

i would like to use your Driver for this Display: 
https://www.dwo.net.cn/pd.jsp?id=11927&fromColId=0#_pp=0_316_38_-1
This page states that there are 2 interface options 3-wire and 4-wire SPI but not QSPI. 

I also found this reference, they are using the lcd-binding-micropython:
https://github.com/Xinyuan-LilyGO/lilygo-micropython/blob/master/target/esp32s3/boards/LILYGO_T-DisplayS3-AMOLED/modules/tft_config.py

There it seems there is an option for SPI and QSPI but i'm not sure if this works.

Is it possible to use your driver with this 3-Wire or 4-Wire SPI instead of QSPI?

Thanks & Best regards.

**Comment by nspsck** [comment]
Hi,

Short answer: sorry, probably not.

Long answer: maybe, but I cannot guarantee it.

This driver is only designed to work with specific lilygo's AMOLED devices which exactly uses an ESP32 and the qspi interface. While it basically overloads the spi implementation from Micropython, the initialization of the bus implemented by ESPRESSIF is experimental and different from Micropython's original spi implementation. So I do not expect it magically to work on a spi interface.

**However**, you can try this: [https://github.com/nspsck/st7735s_WeAct_Studio_TFT_port](https://github.com/nspsck/st7735s_WeAct_Studio_TFT_port)

Even tho the driver is designed for the st7735, it can be easily modified to work on every other display that uses the spi interface and adhire some naming conventions for their registers. Actually it might just work out of the box, because I have seen a lot similarities between these 2 displays during the time writing this driver. But I highly recommend you to modify the `st7789.h`  according to the `rm67162.h`. It then should work 90% assuming the "backlight" part works. (Actually that is probably the only thing that may fail.)

**Comment by gampam2000** [comment]
Hi, I successfully compiled your st7735 driver for micropython esp32. 

But I did not succeed in displaying anything so far.
What I found out is:

- This OLED Display Board does not have an Backlight pin. (do OLEDs have Backlight at all?)
- The Documentation also mentions 3-Wire SPI or 4-Wire SPI, which should I use? How should I connect it?
- https://www.dwo.net.cn/pd.jsp?id=11927&fromColId=0#_pp=0_316_38_-1

What did you mean by modify the st7789.h? I saw that the color definitions are different. Anything else?

Is there away to check if the spi connection is working?

Any help to get this Display working is appreciated ;-) 

Thanks & best regards.

**Comment by nspsck** [comment]
Hi, 

> This OLED Display Board does not have a Backlight pin. (do OLEDs have Backlight at all?)

You are right! The AMOLED does not have a backlight, the function name is just a convention. I was hoping that the backlight on the AMOLED would be turned on during the default initialization path, so you do not have to adjust it anymore. But turned out this was not the case. 

What I suggest you do is modify the lines about brightness according to [these sections](https://github.com/nspsck/RM67162_Micropython_QSPI/blob/86cf3119dd2dd1038e40bccffdb6618a5239d6a3/rm67162/rm67162.c#L1020-L1063).

> The Documentation also mentions 3-Wire SPI or 4-Wire SPI, which should I use? How should I connect it?

![image](https://github.com/nspsck/RM67162_Micropython_QSPI/assets/86347287/7cf5897f-acea-4915-a0c8-3db95a37beff)

If you take a closer look, you only have 2 pins labeled with actual SPI-related functions `(SPI_CLK, SPI_DIN).` I do not consider the `CS`-pin as a SPI-related function, since it's only to tell the client/slave side (in this case, your display) if they should be listening to the master. So this is only a 2-wire-SPI to me (But the book would say that is a 3-wire-SPI, because of the `CS` Pin). The `SPI_D/C` is just straight-up mislabeled and confusing. This is a Data/Command switch on the display side as well, it tells the display, whether the incoming bytes are Data or Commands. 

TL;DR: (from display to your esp32) `SPI_CLK `to `SPI_CLOCK`, `SPI_DIN `to `SPI_MOSI`, `SPI_D/C` to `DC`, `CS `to `CS`. Whereas `SPI_CLOCK` and `SPI_MOSI` are some predefined pins on your esp32 (and not the esp32s3, because it has a MUX, so basically it has more choices), and `DC` and `CS` are GPIO-pins by your choice. 

> What did you mean by modify the st7789.h?

Basically, you compare [st7789.h](https://github.com/nspsck/st7735s_WeAct_Studio_TFT_port/blob/main/st7789/st7789.h) with [rm67162.h](https://github.com/nspsck/RM67162_Micropython_QSPI/blob/main/rm67162/rm67162.h) and check if the commands are the same. Ignore the leading part such as `ST7789_` or 'LCD_CMD_', only the last part matters. You go through the `st7789.h` and check if all the command definitions are the same as in `rm67162.h`. If a command is not present in `st7789.h`, you just ignore it. 

> I saw that the color definitions are different. Anything else?

For example: `INVOFF` and `INVON` are different as well. The rest should be identical, even the `MADCTL_XX` parts. (If you let it to evaluated, it will be the same result). Those displays have really similar command registers, maybe due to industrial needs. 

> Is there a way to check if the spi connection is working?

Use a st7789/st7735/ili9341display to test if you got one. Or use another esp32 as a salve device (technically possible, I just haven't tried it out).

Hopefully, this helps! I sadly do not have one of such displays, hence I can not just bake a spi-driver for it. All I can do is to wish you the best of luck!

Best regards.

**Comment by nspsck** [comment]
And you most likely have to follow the init instructions provided by the lilygo repo you pointed to earlier.

**Comment by gampam2000** [comment]
Hi, thanks for your help.

I modified the [st7789.h](https://github.com/nspsck/st7735s_WeAct_Studio_TFT_port/blob/main/st7789/st7789.h)
I checked/modified all the defines the only difference where those you mentioned. And I added these two:
#define LCD_CMD_WRDISBV      0x51 // Write display brightness
#define LCD_CMD_RDDISBV      0x52 // Read display brightness value

I than modified the [st7789.c](https://github.com/nspsck/st7735s_WeAct_Studio_TFT_port/blob/main/st7789/st7789.c)

```
//modified for rm67162 display
#define backlight_on()                                      \
    {                                                       \
        write_cmd(self, LCD_CMD_WRDISBV, (uint8_t[]) {      \
            0XFF                                            \
        }, 1);                                              \
    }            

#define backlight_off()                                     \
    {                                                       \
        write_cmd(self, LCD_CMD_WRDISBV, (uint8_t[]) {      \
            0X00                                            \
        }, 1);                                              \
    }    
```      

It did compile, but still no luck with displaying anything. I triple checked the connections. Thanks for your explanation of the Pins.


I initialized it like this on an ESP32-S3:
```
from machine import Pin, SPI
import st7789


#Onboard LED on GPIO37:
ob_led = Pin(37,mode=Pin.OUT)

ob_led.on()

rm67162_spi_init = (
    (0xFE, b'\x00', 0),   # PAGE
    (0x36, b'\x00', 0),   # Scan Direction Control
    (0x3A, b'\x75', 0),   # Interface Pixel Format 16bit/pixel
    (0x51, b'\x00', 0),   # Write Display Brightness MAX_VAL=0XFF
    (0x11, b'\x00', 120), # Sleep Out
    (0x29, b'\x00', 120), # Display on
    (0x51, b'\xD0', 0),   # Write Display Brightness MAX_VAL=0XFF
)


tft = st7789.ST7789(
    SPI(1, baudrate=30000000, sck=Pin(39), mosi=Pin(40), miso=None),
    240,
    536,
    reset=Pin(46, Pin.OUT),
    cs=Pin(41, Pin.OUT),
    dc=Pin(45, Pin.OUT),
    backlight=Pin(35, Pin.OUT),
    rotation=0,
    options=0,
    custom_init = rm67162_spi_init,
    #use_drawbuffer=True,
    reversed_backlight=False)

tft.fill(st7789.YELLOW)

ob_led.off()                
```                        

Did I miss something? 
Is there a command to read some register from the display to check if the SPI communication is working?

Thanks & best regards

**Comment by nspsck** [comment]
So,  probably, it is due to the fact, that no backlight pin was provided. During the [initial sequence](https://github.com/nspsck/st7735s_WeAct_Studio_TFT_port/blob/83cafaf0bfc0667fa73900aed2c129b5d05723cb/st7789/st7789.c#L1390-L1439), the "backlight" should be turned on, only if a backlight pin exists.
```c
    if (self->backlight) {
        backlight_on();
    }
```
This is also the case for the `on()` and `off` functions. Initially, this was necessary because the old code actually modifies a backlight Pin. Now you just sending commands, so the check is unwanted and causing issues.

**Comment by nspsck** [comment]
I suggest you assign a random pin to backlight first to check if the display works. (or the driver)

**Comment by gampam2000** [comment]
Hi, thanks for your help. I assigned an unused pin to the backlight, but this didn't help.

I again modified the [st7789.c](https://github.com/nspsck/st7735s_WeAct_Studio_TFT_port/blob/main/st7789/st7789.c):
```

//modified for rm67162 display
#define backlight_on()                                      \
    {                                                       \
        mp_hal_pin_write(self->backlight, 1);               \
        write_cmd(self, LCD_CMD_WRDISBV, (uint8_t[]) {      \
            0XFF                                            \
        }, 1);                                              \
    }            

#define backlight_off()                                     \
    {                                                       \
        mp_hal_pin_write(self->backlight, 0);               \
        write_cmd(self, LCD_CMD_WRDISBV, (uint8_t[]) {      \
            0X00                                            \
        }, 1);                                              \
    }                                                       

```

but still no luck. Did I miss something else?

**Comment by nspsck** [comment]
I again checked your code, have you forgotten to call `tft.init()` before calling `tft.fill()`?

**Comment by gampam2000** [comment]
Hi I got it working ;-)
actually it was the custom_init that caused it not to work.

The python code now looks like this:

```
from machine import Pin, SPI
import st7789
import vga2_16x16 as font

#Onboard LED on GPIO37:
ob_led = Pin(37,mode=Pin.OUT)

ob_led.on()

tft = st7789.ST7789(
    SPI(2, baudrate=20000000, sck=Pin(39), mosi=Pin(40), miso=Pin(48)),
    240,
    536,
    reset=Pin(42, Pin.OUT),
    cs=Pin(45, Pin.OUT),
    dc=Pin(46, Pin.OUT),
    backlight=Pin(47, Pin.OUT),
    rotations=[(0x00, 240, 536, 0, 0), (0x60, 536, 240, 0, 0), (0xc0, 240, 536, 0, 0), (0xa0, 536, 240, 0, 0)],
    rotation=1,
    options=0,
    inversion=False,
    color_order=st7789.RGB,
    use_drawbuffer=True,
    reversed_backlight=False)

tft.init()
tft.fill(st7789.YELLOW)

text="Hallo"
length = len(text)

tft.text(
        font,
        text,
        tft.width() // 2 - length // 2 * font.WIDTH,
        tft.height() // 2 - font.HEIGHT,
        st7789.BLACK,
        st7789.YELLOW)
```

It seems that the color definitions are wrong (eg YELLOW is displayed GREEN, Black is White,...) 
I tried all combinations of inversion and color_order with no luck.

I took the definitions from here:
https://github.com/nspsck/RM67162_Micropython_QSPI/blob/86cf3119dd2dd1038e40bccffdb6618a5239d6a3/rm67162/rm67162.h#L60C1-L73C1

Why could this be?

**Comment by nspsck** [comment]
Hi,
>Hi I got it working ;-)
actually it was the custom_init that caused it not to work.

Great! Good job!

>Why could this be? 

Please try setting a parameter called `inversion` to `True`. This should solve your issue. What it effectively does is:

Some displays have a inv bit of `0x21` and the others have it for `0x20`, as above I have already mentioned. 
> For example: INVOFF and INVON are different as well. The rest should be identical, even the MADCTL_XX parts. (If you let it to evaluated, it will be the same result). Those displays have really similar command registers, maybe due to industrial needs.

Because this is almost arbitrarily chosen by the manufacturer, there is no trivial hardcoded solution for this problem. So you have to set it manually.

**Comment by gampam2000** [comment]
Hi, 

i recompiled it with the original color definitions of  https://github.com/nspsck/st7735s_WeAct_Studio_TFT_port/blob/83cafaf0bfc0667fa73900aed2c129b5d05723cb/st7789/st7789.h#L55C1-L63C23

and now the colors are correct. Thanks for your driver and help. Maybe i will create a fork of your driver for the rm67162.

**Comment by gampam2000** [comment]
Hi, I created  a fork for the SPI mode:
https://github.com/gampam2000/RM67162_Micropyton_SPI
If you wish you can reference it in your documentation.
Thanks again for your Help & Best regards.

**Comment by nspsck** [comment]
Hi, certainly. And thank you for you work!

Ugh... but can you please keep the credits to [russhughes](https://github.com/russhughes/st7789_mpy/tree/master/st7789)? He is the original creator. All I did was some modifications/optimizations. 

Best Regards

**Comment by gampam2000** [comment]
Hi, no problem I updated the Readme and included russ and you.
Best regards
